zcash_client_sqlite/src/wallet/scanning.rs

```diff
@@ -0,0 +1,1107 @@
+    use ScanPriority::*;
+    match code {
+        10 => Some(Scanned),
+        20 => Some(Historic),
+        30 => Some(OpenAdjacent),
+        40 => Some(FoundNote),
+        50 => Some(ChainTip),
+        60 => Some(Verify),
+        _ => None,
+    }
+}
+
+pub(crate) fn priority_code(priority: &ScanPriority) -> i64 {
+    use ScanPriority::*;
+    match priority {
+        Scanned => 10,
+        Historic => 20,
+        OpenAdjacent => 30,
+        FoundNote => 40,
+        ChainTip => 50,
+        Verify => 60,
+    }
+}
+
+pub(crate) fn suggest_scan_ranges(
+    conn: &rusqlite::Connection,
+    min_priority: Option<ScanPriority>,
+) -> Result<Vec<ScanRange>, SqliteClientError> {
+    let mut stmt_scan_ranges = conn.prepare_cached(
+        "SELECT block_range_start, block_range_end, priority
+         FROM scan_queue
+         WHERE priority >= :min_priority
+         ORDER BY priority DESC, block_range_end DESC",
+    )?;
+
+    let mut rows = stmt_scan_ranges.query(named_params![
+        ":min_priority": priority_code(&min_priority.unwrap_or(ScanPriority::Historic))
+    ])?;
+
+    let mut result = vec![];
+    while let Some(row) = rows.next()? {
+        let range = Range {
+            start: row.get::<_, u32>(0).map(BlockHeight::from)?,
+            end: row.get::<_, u32>(1).map(BlockHeight::from)?,
+        };
+        let code = row.get::<_, i64>(2)?;
+        let priority = parse_priority_code(code).ok_or_else(|| {
+            SqliteClientError::CorruptedData(format!("scan priority not recognized: {}", code))
+        })?;
+
+        result.push(ScanRange::from_parts(range, priority));
+    }
+
+    Ok(result)
+}
+
+// This implements the dominance rule for range priority. If the inserted range's priority is
+// `Verify`, this replaces any existing priority. Otherwise, if the current priority is
+// `Scanned`, this overwrites any priority
+fn update_priority(current: ScanPriority, inserted: ScanPriority) -> ScanPriority {
+    match (current, inserted) {
+        (_, ScanPriority::Verify) => ScanPriority::Verify,
+        (ScanPriority::Scanned, _) => ScanPriority::Scanned,
+        (_, ScanPriority::Scanned) => ScanPriority::Scanned,
+        (a, b) => max(a, b),
+    }
+}
+
+fn dominance(current: &ScanPriority, inserted: &ScanPriority, insert: Insert) -> Dominance {
+    match (current, inserted) {
+        (_, ScanPriority::Verify | ScanPriority::Scanned) => Dominance::from(insert),
+        (ScanPriority::Scanned, _) => Dominance::from(!insert),
+        (a, b) => match a.cmp(b) {
+            Ordering::Less => Dominance::from(insert),
+            Ordering::Equal => Dominance::Equal,
+            Ordering::Greater => Dominance::from(!insert),
+        },
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum RangeOrdering {
+    LeftFirstDisjoint,
+    LeftFirstOverlap,
+    LeftContained,
+    Equal,
+    RightContained,
+    RightFirstOverlap,
+    RightFirstDisjoint,
```

Comment:
```suggestion
/// In the comments for each alternative, `()` represents the left range and `[]` represents the right range.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum RangeOrdering {
    /// `(   ) [   ]`
    LeftFirstDisjoint,
    /// `( [   ) ]`
    LeftFirstOverlap,
    /// `( [   ] )` CHECK
    LeftContained,
    /// ```
    /// (   )
    /// [   ]
    /// ```
    Equal,
    /// `[ (   ) ]` CHECK
    RightContained,
    /// `[ (   ] )`
    RightFirstOverlap,
    /// `[   ] (   )`
    RightFirstDisjoint,
```